src/model/rpc.js

```diff
@@ -31,44 +30,95 @@ function jsErrorToBtpError (e) {
     assert(typeof plugin === 'object', 'plugin must be provided')
 
     super()
     this._sockets = []
     this._handlers = handlers
-    this._rpcUri = rpcUri
+    this._client = client
     this._plugin = plugin
+    this._incomingAuthToken = incomingAuthToken
     this.debug = debug
   }
 
-  addSocket (socket) {
-    _assertSocket(socket)
+  async addSocket (socket, authToken) {
+    const newSocketIndex = this._sockets.length
+    const isClient = Boolean(authToken)
+    _assertSocket({ socket, authorized: isClient })
+
     this.debug('adding socket')
-    this._sockets.push(socket)
+
+    // if we're the client on this socket, we don't need to receive
+    // any authentication data. we have to send it instead.
+    this._sockets.push({ socket, authorized: isClient })
     socket.on('message', async (message) => {
       this.debug('got message:', Buffer.from(message).toString('hex'))
       try {
-        await this.handleMessage(socket, message)
+        await this.handleMessage(newSocketIndex, message)
       } catch (err) {
         this.debug(`RPC Error: ${err.message}. Message was ${JSON.stringify(message)}`)
       }
     })
+
+    // if this is a client, then send a special request with which to
+    // authenticate.
+    if (isClient) {
+      const requestId = await _requestId()
+      await _send(socket, btpPacket.serializeMessage(requestId, [{
+        protocolName: 'auth',
+        contentType: btpPacket.MIME_APPLICATION_OCTET_STREAM,
+        data: Buffer.from([])
+      }, {
+        protocolName: 'auth_token',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from(authToken, 'utf8')
+      }, {
+        protocolName: 'auth_username',
+        contentType: btpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from('', 'utf8')
+      }]))
+
+      return new Promise((resolve, reject) => {
+        const handleAuthResponse = (type, data) => {
+          if (type === btpPacket.TYPE_RESPONSE) {
+            resolve(data)
+          } else if (type === btpPacket.TYPE_ERROR) {
+            reject(new Error(JSON.stringify(data)))
+          } else {
+            reject(new Error('Unkown BTP packet type', data))
+          }
+        }
+
+        this.once('_' + requestId, handleAuthResponse)
+      })
+    }
   }
 
   setAuthToken (token) {
     this._token = token
   }
 
-  async handleMessage (socket, message) {
-    _assertSocket(socket)
+  async handleMessage (socketIndex, message) {
+    const socketData = this._sockets[socketIndex]
+    _assertSocket(socketData)
+
+    const socket = socketData.socket
     const {type, requestId, data} = btpPacket.deserialize(message)
     const typeString = btpPacket.typeToString(type)
+
+    if (!socketData.authorized) {
+      // authentication handling must be done inside of the RPC module because
+      // it happens on a per-socket basis rather than per plugin.
+      this._handleAuth(socketIndex, { type, requestId, data })
```

Comment:
There should be a timeout for the client to send the auth details.

Reply:
added